src/transaction/lock.rs

```diff
@@ -21,67 +22,74 @@ pub async fn resolve_locks(
+    type LockVersion = u64;
     let mut has_live_locks = false;
-    let expired_locks = locks.into_iter().filter(|lock| {
+
+    struct ResolveInfo {
+        region: Region,
+        primary_key: Key,
+        keys: Vec<Key>,
+    }
+
+    let mut grouped: HashMap<(RegionVerId, LockVersion), ResolveInfo> = HashMap::new();
+
+    for lock in locks {
         let expired = ts.physical - Timestamp::from_version(lock.lock_version).physical
             >= lock.lock_ttl as i64;
-        if !expired {
+        if expired {
+            let primary_key: Key = lock.primary_lock.into();
+            let region: Region = pd_client.region_for_key(&primary_key).await?;
+            let is_small_txn = lock.txn_size < 16;
+
+            let resolve_info = grouped
+                .entry((region.ver_id(), lock.lock_version))
+                .or_insert(ResolveInfo {
+                    region,
+                    primary_key: primary_key.clone(),
+                    keys: vec![],
+                });
+
+            if is_small_txn {
+                resolve_info.keys.push(primary_key);
+            } else {
+                //if txn is large,keep resolve_info.keys empty to send a full resolve lock request
+            }
+
+        } else {
             has_live_locks = true;
         }
-        expired
-    });
-
-    // records the commit version of each primary lock (representing the status of the transaction)
-    let mut commit_versions: HashMap<u64, u64> = HashMap::new();
-    let mut clean_regions: HashMap<u64, HashSet<RegionVerId>> = HashMap::new();
-    for lock in expired_locks {
-        let primary_key: Key = lock.primary_lock.into();
-        let region_ver_id = pd_client.region_for_key(&primary_key).await?.ver_id();
-        // skip if the region is cleaned
-        if clean_regions
-            .get(&lock.lock_version)
-            .map(|regions| regions.contains(&region_ver_id))
-            .unwrap_or(false)
-        {
-            continue;
-        }
+    }
 
-        let commit_version = match commit_versions.get(&lock.lock_version) {
-            Some(&commit_version) => commit_version,
-            None => {
-                let commit_version = requests::new_cleanup_request(primary_key, lock.lock_version)
-                    .execute(pd_client.clone())
-                    .await?;
-                commit_versions.insert(lock.lock_version, commit_version);
-                commit_version
-            }
-        };
+    for ((_region_ver_id, lock_version), resolve_info) in grouped {
+        let commit_version =
+            requests::new_cleanup_request(resolve_info.primary_key.clone(), lock_version)
+                .execute(pd_client.clone())
+                .await?;
 
-        let cleaned_region = resolve_lock_with_retry(
-            lock.key.into(),
-            lock.lock_version,
+        let _cleaned_region = resolve_lock_with_retry(
+            &resolve_info.region,
+            &resolve_info.keys,
```

Comment:
You can change it to take the ownership and move it in to avoid cloning in the later processes.

Reply:
> You can change it to take the ownership and move it in to avoid cloning in the later processes.

cloning in the later process seems can't be reduced since clone happens in a loop.

Reply:
If you just take a `Vec<Key>`, I think we can write `req.set_keys(keys.into_iter().map(Into::into).collect());` [here](https://github.com/tikv/client-rust/pull/145/files#diff-d01850033b0d2fc084df1829a7789724R264)

Reply:
> If you just take a `Vec<Key>`, I think we can write `req.set_keys(keys.into_iter().map(Into::into).collect());` [here](https://github.com/tikv/client-rust/pull/145/files#diff-d01850033b0d2fc084df1829a7789724R264)

If `new_resolve_lock_request` take a `Vec<Key>`, `value moved here, in previous iteration of loop` will be generated at https://github.com/tikv/client-rust/pull/145/files#diff-49f989cfce4dadfa41d7d1bf2548edd7R99

Reply:
Ah sorry, I got what's the limitation here.